File: examples/wallet_eip1193/guard.ts
/**
 * e_AI v2 Wallet Guard — EIP-1193 provider wrapper.
 *
 * STATUS: ILLUSTRATIVE ADAPTER, NOT a profile-driven runtime. The
 * `profiles` parameter is accepted but the alert logic in `analyzeTx` /
 * `analyzeSignature` currently uses hard-coded selectors, thresholds,
 * and alert metadata — it does NOT dispatch from profile semantics.
 * A canonical profile-driven runtime is queued as a maturity-gate
 * item; until then, treat this file as a pattern reference for how a
 * wallet wrapper would integrate, not as the canonical wallet
 * runtime. See README "What's available — Integration demos (status:
 * illustrative adapters)" section.
 *
 * Works with ANY wallet that implements EIP-1193 (MetaMask, Rabby, Frame,
 * WalletConnect, injected providers, etc.)
 *
 * Intercepts:
 *   - eth_sendTransaction → approval_phishing, governance_proposal, l2_bridge_linkage
 *   - eth_signTypedData_v4 → offchain_signature
 *   - eth_call / eth_getBalance → rpc_leakage
 *
 * Usage:
 *   const guarded = withEIP1193Guard(window.ethereum, profiles);
 *   // Now use `guarded` instead of `window.ethereum` everywhere
 */

// --- Types ---

interface EIP1193Provider {
  request(args: { method: string; params?: unknown[] }): Promise<unknown>;
  on?(event: string, handler: (...args: unknown[]) => void): void;
}

interface Profile {
  meta: { domain_name: string };
  heuristics: Record<string, Heuristic>;
  [key: string]: unknown;
}

interface Heuristic {
  id: string;
  name: string;
  severity: string;
  description: string;
  detection: { signals: Signal[] };
  recommendations: Recommendation[];
}

interface Signal {
  name: string;
  description: string;
  confidence: number;
}

interface Recommendation {
  action: string;
  description: string;
  effectiveness: number;
}

interface Alert {
  profile: string;
  heuristic: string;
  severity: string;
  confidence: number;
  signal: string;
  recommendation: string;
}

interface GuardConfig {
  profiles: Profile[];
  onAlert?: (alerts: Alert[], action: 'block' | 'warn' | 'pass') => void | Promise<void>;
  onBlock?: (alerts: Alert[]) => void | Promise<void>;
  blockOnCritical?: boolean;  // default: true
  trackRpcPatterns?: boolean;  // default: true
}

// --- Known selectors ---

const SELECTORS: Record<string, { name: string; profiles: string[] }> = {
  '0x095ea7b3': { name: 'approve', profiles: ['approval_phishing'] },
  '0x39509351': { name: 'increaseAllowance', profiles: ['approval_phishing'] },
  '0xa22cb465': { name: 'setApprovalForAll', profiles: ['approval_phishing', 'offchain_signature'] },
  '0xd505accf': { name: 'permit', profiles: ['offchain_signature'] },
  '0xfe9fbb80': { name: 'execute', profiles: ['governance_proposal'] },
  '0xda95691a': { name: 'propose', profiles: ['governance_proposal'] },
  '0x56781388': { name: 'castVote', profiles: ['governance_proposal'] },
  // Bridge selectors vary by bridge contract -- would be loaded from profile
};

const MAX_UINT256 = BigInt('0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff');

// --- RPC Pattern Tracker ---

class RpcPatternTracker {
  private queries: { method: string; params: string; timestamp: number }[] = [];
  private readonly windowMs = 5 * 60 * 1000; // 5 min window

  record(method: string, params: unknown[]): void {
    this.queries.push({
      method,
      params: JSON.stringify(params).slice(0, 200),
      timestamp: Date.now(),
    });
    // Prune old
    const cutoff = Date.now() - this.windowMs;
    this.queries = this.queries.filter(q => q.timestamp > cutoff);
  }

  analyze(): Alert[] {
    const alerts: Alert[] = [];
    const methods = this.queries.map(q => q.method);

    // H1: Multiple getBalance calls linking addresses
    const balanceCalls = this.queries.filter(q => q.method === 'eth_getBalance');
    const uniqueAddresses = new Set(balanceCalls.map(q => {
      try { return JSON.parse(q.params)[0]; } catch { return ''; }
    }));
    if (uniqueAddresses.size > 3) {
      alerts.push({
        profile: 'rpc_leakage',
        heuristic: 'H1',
        severity: 'high',
        confidence: 0.8,
        signal: `Balance checked for ${uniqueAddresses.size} addresses in ${this.windowMs / 60000} min — RPC provider now knows these addresses are related`,
        recommendation: 'Use Helios local light client or batch queries through Tor',
      });
    }

    // H2: Repeated eth_call (position monitoring)
    const callCount = methods.filter(m => m === 'eth_call').length;
    if (callCount > 10) {
      alerts.push({
        profile: 'rpc_leakage',
        heuristic: 'H2',
        severity: 'medium',
        confidence: 0.6,
        signal: `${callCount} eth_call queries in ${this.windowMs / 60000} min — pattern reveals DeFi position monitoring`,
        recommendation: 'Reduce polling frequency or use local node',
      });
    }

    return alerts;
  }
}

// --- Transaction Analyzer ---

// --- ABI decoding helpers ---

const HEX_RE = /^0x[0-9a-fA-F]*$/;

function isWellFormedCalldata(data: string, expectedBytes: number): boolean {
  // expectedBytes counts the bytes AFTER the 4-byte selector.
  // Total length = 2 (for "0x") + 8 (selector) + 2*expectedBytes.
  if (!HEX_RE.test(data)) return false;
  const expectedLen = 2 + 8 + 2 * expectedBytes;
  return data.length === expectedLen;
}

function decodeAddress(data: string, byteOffset: number): string {
  // Solidity ABI: address is right-aligned in a 32-byte word. The address
  // bytes are at offset byteOffset+12 .. byteOffset+32 (skipping the 12-byte
  // left-pad). String slice in hex: skip "0x" (2 chars) + 2*byteOffset + 24
  // chars of left-pad = 2 + 2*byteOffset + 24, then take 40 chars (20 bytes
  // × 2 hex chars).
  const start = 2 + 2 * byteOffset + 24;
  return '0x' + data.slice(start, start + 40);
}

function decodeUint256(data: string, byteOffset: number): bigint {
  const start = 2 + 2 * byteOffset;
  const hex = '0x' + data.slice(start, start + 64);
  return BigInt(hex);
}

function decodeBool(data: string, byteOffset: number): boolean {
  // Solidity ABI: bool is encoded in a 32-byte word; non-zero word = true.
  // We read as uint256 and compare against zero.
  return decodeUint256(data, byteOffset) !== BigInt(0);
}

function analyzeTx(to: string, data: string, value: bigint, profiles: Profile[]): Alert[] {
  const alerts: Alert[] = [];
  if (!data || data.length < 10) return alerts;

  const selector = data.slice(0, 10);
  const known = SELECTORS[selector];

  // Approval checks: approve(address spender, uint256 amount)
  // Calldata layout: selector (4B) + spender (32B word, right-aligned addr) + amount (32B)
  if (selector === '0x095ea7b3' || selector === '0x39509351') {
    if (!isWellFormedCalldata(data, 64)) {
      // Malformed approval calldata — fail closed with a high-severity alert
      alerts.push({
        profile: 'approval_phishing',
        heuristic: 'malformed',
        severity: 'high',
        confidence: 0.9,
        signal: `Malformed approve() calldata (length ${data.length}, expected 138 hex chars). Refusing to silently pass.`,
        recommendation: 'Reject the transaction; calldata cannot be decoded. Most likely a wallet/dApp bug or spoof attempt.',
      });
      return alerts;
    }
    let spender: string;
    let amount: bigint;
    try {
      spender = decodeAddress(data, 4);
      amount = decodeUint256(data, 4 + 32);
    } catch (e) {
      alerts.push({
        profile: 'approval_phishing',
        heuristic: 'decode_error',
        severity: 'high',
        confidence: 0.8,
        signal: `Could not decode approve() calldata: ${(e as Error).message}`,
        recommendation: 'Reject the transaction.',
      });
      return alerts;
    }

    if (amount >= MAX_UINT256 / BigInt(2)) {
      alerts.push({
        profile: 'approval_phishing',
        heuristic: 'H1',
        severity: 'high',
        confidence: 0.95,
        signal: `Unlimited approval — token=${to}, spender=${spender}`,
        recommendation: 'Set approval to exact amount needed, not unlimited. Verify the spender address.',
      });
    }
  }

  // setApprovalForAll(address operator, bool approved)
  // Calldata layout: selector (4B) + operator (32B word) + approved (32B word)
  if (selector === '0xa22cb465') {
    if (!isWellFormedCalldata(data, 64)) {
      alerts.push({
        profile: 'approval_phishing',
        heuristic: 'malformed',
        severity: 'high',
        confidence: 0.9,
        signal: `Malformed setApprovalForAll() calldata (length ${data.length}). Refusing to silently pass.`,
        recommendation: 'Reject the transaction.',
      });
      return alerts;
    }
    let operator: string;
    let approved: boolean;
    try {
      operator = decodeAddress(data, 4);
      approved = decodeBool(data, 4 + 32);
    } catch {
      operator = '<decode-failed>';
      approved = true;  // fail closed: alert if we can't tell
    }

    // Codex Phase 3 review missed-coverage #2: setApprovalForAll(false) is
    // a REVOCATION, not a grant — emit an info-level note rather than the
    // high-severity grant alert. Previously fired the same alert for both.
    if (!approved) {
      alerts.push({
        profile: 'approval_phishing',
        heuristic: 'H4_revoke',
        severity: 'low',
        confidence: 0.9,
        signal: `setApprovalForAll(false) — REVOKING operator approval, collection=${to}, operator=${operator}`,
        recommendation: 'No action needed; revocation is a defensive operation.',
      });
    } else {
      alerts.push({
        profile: 'approval_phishing',
        heuristic: 'H4',
        severity: 'high',
        confidence: 0.75,
        signal: `setApprovalForAll(true) — collection=${to}, operator=${operator}`,
        recommendation: 'Verify the operator contract is legitimate before approving all NFTs',
      });
    }
  }

  return alerts;
}

// --- Signature Analyzer ---

function analyzeSignature(typedData: unknown): Alert[] {
  const alerts: Alert[] = [];
  if (!typedData || typeof typedData !== 'object') return alerts;

  const td = typedData as Record<string, unknown>;
  const primaryType = td.primaryType as string || '';
  const domain = td.domain as Record<string, unknown> || {};
  const message = td.message as Record<string, unknown> || {};

  // Permit2
  if (primaryType === 'PermitSingle' || primaryType === 'PermitBatch') {
    const amount = message.details && typeof message.details === 'object'
      ? (message.details as Record<string, unknown>).amount
      : message.amount;

    alerts.push({
      profile: 'offchain_signature',
      heuristic: primaryType === 'PermitBatch' ? 'H5' : 'H1',
      severity: 'high',
      confidence: 0.9,
      signal: `${primaryType} signature — authorizes token transfer without on-chain approval tx`,
      recommendation: 'Check the spender address and amount carefully. Set short expiration.',
    });
  }

  // Seaport order
  if (primaryType === 'OrderComponents' || domain.name === 'Seaport') {
    alerts.push({
      profile: 'offchain_signature',
      heuristic: 'H6',
      severity: 'medium',
      confidence: 0.7,
      signal: 'Seaport order signature — verify the listing price matches your intent',
      recommendation: 'Check that offer and consideration items match expected values',
    });
  }

  // Generic EIP-712 with transfer-like fields
  if (message.to && message.value) {
    alerts.push({
      profile: 'offchain_signature',
      heuristic: 'H2',
      severity: 'high',
      confidence: 0.8,
      signal: 'EIP-712 typed data contains transfer fields (to, value) — this signature may authorize a token transfer',
      recommendation: 'Verify this is not a disguised approval or transfer authorization',
    });
  }

  return alerts;
}

// --- Main Guard ---

export function withEIP1193Guard(
  provider: EIP1193Provider,
  config: GuardConfig,
): EIP1193Provider {
  const rpcTracker = config.trackRpcPatterns !== false ? new RpcPatternTracker() : null;
  const blockOnCritical = config.blockOnCritical !== false;

  return {
    ...provider,

    async request(args: { method: string; params?: unknown[] }): Promise<unknown> {
      const { method, params = [] } = args;
      let alerts: Alert[] = [];

      // --- Layer 3: Transaction Guard ---
      if (method === 'eth_sendTransaction') {
        const tx = params[0] as Record<string, string>;
        const to = tx?.to || '';
        const data = tx?.data || '0x';
        const value = tx?.value ? BigInt(tx.value) : BigInt(0);

        alerts = analyzeTx(to, data, value, config.profiles);
      }

      // --- Layer 3: Signature Guard ---
      if (method === 'eth_signTypedData_v4') {
        const typedDataStr = params[1] as string;
        try {
          const typedData = JSON.parse(typedDataStr);
          alerts = analyzeSignature(typedData);
        } catch {}
      }

      // --- Layer 2: Provider Guard (RPC tracking) ---
      if (rpcTracker) {
        rpcTracker.record(method, params);

        if (method === 'eth_getBalance' || method === 'eth_call') {
          const rpcAlerts = rpcTracker.analyze();
          alerts.push(...rpcAlerts);
        }
      }

      // --- Dispatch alerts ---
      if (alerts.length > 0) {
        const hasCritical = alerts.some(a => a.severity === 'critical' && a.confidence > 0.8);
        const action = hasCritical && blockOnCritical ? 'block' : alerts.length > 0 ? 'warn' : 'pass';

        if (config.onAlert) {
          await config.onAlert(alerts, action);
        }

        if (action === 'block') {
          if (config.onBlock) {
            await config.onBlock(alerts);
          }
          throw new Error(
            `e_AI Guard blocked ${method}: ${alerts.map(a => `[${a.heuristic}] ${a.signal}`).join('; ')}`
          );
        }
      }

      // --- Pass through ---
      return provider.request(args);
    },
  };
}


// --- Demo ---

async function demo() {
  // Mock provider
  const mockProvider: EIP1193Provider = {
    async request({ method, params }) {
      console.log(`  [Provider] ${method}(${JSON.stringify(params).slice(0, 80)}...)`);
      if (method === 'eth_sendTransaction') return '0xtxhash';
      if (method === 'eth_getBalance') return '0x1000000000000000';
      return null;
    },
  };

  const guarded = withEIP1193Guard(mockProvider, {
    profiles: [],
    blockOnCritical: true,
    onAlert: (alerts, action) => {
      console.log(`\n  [Guard] ${action.toUpperCase()}: ${alerts.length} alert(s)`);
      for (const a of alerts) {
        console.log(`    [${a.profile}/${a.heuristic}] ${a.severity} (${(a.confidence * 100).toFixed(0)}%): ${a.signal}`);
        console.log(`    → ${a.recommendation}`);
      }
    },
    onBlock: (alerts) => {
      console.log(`\n  ⛔ BLOCKED`);
    },
  });

  console.log('=== Scenario 1: Unlimited approval ===');
  try {
    await guarded.request({
      method: 'eth_sendTransaction',
      params: [{
        to: '0x1234567890abcdef1234567890abcdef12345678',
        data: '0x095ea7b3000000000000000000000000spender0000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
        value: '0x0',
      }],
    });
  } catch (e: any) { console.log(`  Caught: ${e.message.slice(0, 80)}`); }

  console.log('\n=== Scenario 2: Permit2 signature ===');
  try {
    await guarded.request({
      method: 'eth_signTypedData_v4',
      params: [
        '0xuser',
        JSON.stringify({
          primaryType: 'PermitSingle',
          domain: { name: 'Permit2' },
          message: { details: { amount: '115792089237316195423570985008687907853269984665640564039457584007913129639935' } },
        }),
      ],
    });
  } catch (e: any) { console.log(`  Caught: ${e.message.slice(0, 80)}`); }

  console.log('\n=== Scenario 3: Multiple balance checks (RPC leakage) ===');
  for (const addr of ['0xaaa', '0xbbb', '0xccc', '0xddd']) {
    await guarded.request({ method: 'eth_getBalance', params: [addr, 'latest'] });
  }

  console.log('\n=== Done ===');
}

// Run if executed directly
const isMain = typeof process !== 'undefined' && process.argv[1]?.endsWith('guard.ts');
if (isMain) demo().catch(console.error);

File: examples/ai_agent/guard.py
"""
e_AI v2 AI Agent Guard — intercepts agent tool calls before execution.

STATUS: ILLUSTRATIVE ADAPTER, NOT a profile-driven runtime.
Per Codex 2026-05-07 review, the demo's two advertised risky
scenarios ("Agent approves unlimited tokens" and "Agent checks 5
wallet balances") currently DO NOT emit warnings — the decorator
maps approval-tool intent to `eth_sendTransaction` but only passes
function metadata, not calldata; balance-check tool maps to
`eth_call` but the rpc_leakage detector counts only `eth_getBalance`.
Plus the loader globs `domains/*/profile.json` indiscriminately,
including v1 variants (loads 22 profiles, not 16). Hardening + tool-
intent adapter wiring is queued. Treat this file as a pattern
reference for how an agent guard would integrate, not as the
canonical agent runtime. See README "Integration demos (status:
illustrative adapters)" section.

Works with any agent framework (LangChain, OpenAI function calling,
custom agents). Wraps tool execution with profile-based risk analysis.

For AI access method: the agent queries on-chain data and submits
transactions on behalf of the user. The guard checks both:
  1. Query patterns (rpc_leakage profile) — does the query reveal strategy?
  2. Transaction intent (approval_phishing, etc.) — is the action safe?

Usage:
    from examples.ai_agent.guard import AgentGuard

    guard = AgentGuard(profiles_dir="domains/")

    # Wrap tool execution
    @guard.check
    def execute_swap(token_in, token_out, amount):
        ...

    # Or check manually
    alerts = guard.check_action("eth_sendTransaction", {...})
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable, Any


@dataclass
class Alert:
    profile: str
    heuristic: str
    severity: str
    confidence: float
    signal: str
    recommendation: str


@dataclass
class GuardResult:
    action: str  # "pass" | "warn" | "block"
    alerts: list[Alert] = field(default_factory=list)


class AgentGuard:
    """Guards AI agent actions against e_AI v2 profiles."""

    def __init__(
        self,
        profiles_dir: str = "domains/",
        block_on_critical: bool = True,
    ):
        self.profiles = self._load_profiles(profiles_dir)
        self.block_on_critical = block_on_critical
        self.rpc_log: list[dict] = []

    # Skip set: _template + _feedback + 6 v1 sanitization variants (Part 3
    # supporting material, not v2 production guards). Per docs/v1_variants.md.
    _SKIP_DOMAINS = {
        "_template",
        "_feedback",
        "defi",
        "defi_14b",
        "defi_bootstrap",
        "defi_claude",
        "defi_generated",
        "defi_websearch",
    }

    def _load_profiles(self, profiles_dir: str) -> dict[str, dict]:
        """Load v2 production profile.json files from domain directories."""
        profiles = {}
        base = Path(profiles_dir)
        for domain_dir in sorted(base.iterdir()):
            if not domain_dir.is_dir():
                continue
            if domain_dir.name in self._SKIP_DOMAINS:
                continue
            profile_path = domain_dir / "profile.json"
            if profile_path.exists():
                with open(profile_path) as f:
                    profiles[domain_dir.name] = json.load(f)
        return profiles

    def check_action(self, method: str, params: dict) -> GuardResult:
        """Check an agent action against all relevant profiles.

        Args:
            method: "eth_sendTransaction", "eth_signTypedData", "eth_call", etc.
            params: Method-specific parameters

        Returns:
            GuardResult with action (pass/warn/block) and alerts
        """
        alerts = []

        # Transaction checks
        if method == "eth_sendTransaction":
            alerts.extend(self._check_tx(params))

        # Signature checks
        if method in ("eth_signTypedData_v4", "eth_signTypedData"):
            alerts.extend(self._check_signature(params))

        # RPC pattern checks
        if method in ("eth_getBalance", "eth_call", "eth_getLogs"):
            self.rpc_log.append({"method": method, "params": params, "time": time.time()})
            alerts.extend(self._check_rpc_pattern())

        # Determine action
        has_critical = any(a.severity == "critical" and a.confidence > 0.8 for a in alerts)
        if has_critical and self.block_on_critical:
            action = "block"
        elif alerts:
            action = "warn"
        else:
            action = "pass"

        return GuardResult(action=action, alerts=alerts)

    # ABI selectors for tool-intent → calldata adaptation
    _APPROVE_SELECTOR = "0x095ea7b3"  # approve(address,uint256)
    _MAX_UINT256_HEX = "f" * 64

    def _build_tool_intent(self, fn_name: str, args: tuple, kwargs: dict) -> tuple[str, dict]:
        """Translate an agent tool call into (rpc_method, params) the guard
        can analyze. Adapter — tool intent → RPC-shaped params.

        Specifically handles:
          - approval tools: build calldata so _check_tx sees a real
            unlimited-approval pattern, not just function metadata
          - balance tools: emit eth_getBalance with the queried address
            so _check_rpc_pattern can count balance-checked addresses
          - signing tools: pass typed-data through
          - other transaction tools: emit eth_sendTransaction with
            best-effort target inference
        """
        fn_lower = fn_name.lower()

        # Approval tools: approve_token(token, spender, amount)
        if "approve" in fn_lower and len(args) >= 2:
            token = str(args[0]) if len(args) > 0 else kwargs.get("token", "0x")
            spender = str(args[1]) if len(args) > 1 else kwargs.get("spender", "0x")
            amount = args[2] if len(args) > 2 else kwargs.get("amount", "")
            # Build calldata: selector + spender (32B left-pad) + amount (32B left-pad)
            spender_hex = spender.lower().replace("0x", "").rjust(64, "0")
            if isinstance(amount, str) and amount.lower() == "unlimited":
                amount_hex = self._MAX_UINT256_HEX
            else:
                try:
                    amt_int = int(amount) if not isinstance(amount, int) else amount
                    amount_hex = format(amt_int, "x").rjust(64, "0")
                except (TypeError, ValueError):
                    amount_hex = self._MAX_UINT256_HEX  # be conservative
            data = self._APPROVE_SELECTOR + spender_hex + amount_hex
            return "eth_sendTransaction", {"to": token, "data": data, "_tool": fn_name}

        # Balance tools: check_balance(address) → eth_getBalance, NOT eth_call
        # (the rpc_leakage pattern detector counts eth_getBalance specifically).
        if any(w in fn_lower for w in ["balance", "get_balance"]):
            address = str(args[0]) if args else kwargs.get("address", "")
            return "eth_getBalance", [address, "latest"]

        # Position / price tools → eth_call against an address
        if any(w in fn_lower for w in ["price", "position"]):
            target = str(args[0]) if args else kwargs.get("target", "")
            return "eth_call", [{"to": target, "data": "0x"}, "latest"]

        # Signing tools
        if any(w in fn_lower for w in ["sign", "permit"]):
            typed = args[0] if args and isinstance(args[0], dict) else kwargs.get("typed_data", {})
            return "eth_signTypedData_v4", {"typed_data": typed}

        # Generic transaction tools (swap/transfer/send/stake/deposit/withdraw)
        if any(w in fn_lower for w in ["swap", "transfer", "send", "stake", "deposit", "withdraw"]):
            target = str(args[0]) if args else kwargs.get("to", "")
            return "eth_sendTransaction", {"to": target, "data": "0x", "_tool": fn_name}

        return "unknown", {"function": fn_name, "args": args, "kwargs": kwargs}

    def check(self, fn: Callable) -> Callable:
        """Decorator that checks agent tool calls before execution.

        The decorator translates tool intent into RPC-shaped params so the
        underlying analyzers (`_check_tx`, `_check_rpc_pattern`,
        `_check_signature`) see realistic input. Previously the wrapper
        passed only function metadata (name + args), which silently
        bypassed every detector.
        """
        def wrapper(*args, **kwargs):
            fn_name = fn.__name__
            method, params = self._build_tool_intent(fn_name, args, kwargs)

            result = self.check_action(method, params)

            if result.action == "block":
                raise RuntimeError(
                    f"AgentGuard BLOCKED {fn_name}: "
                    + "; ".join(f"[{a.heuristic}] {a.signal}" for a in result.alerts)
                )

            if result.action == "warn":
                print(f"  [AgentGuard] WARNING for {fn_name}:")
                for a in result.alerts:
                    print(f"    [{a.profile}/{a.heuristic}] {a.signal}")

            return fn(*args, **kwargs)
        return wrapper

    def _check_tx(self, params: dict) -> list[Alert]:
        """Check transaction against approval/governance/bridge profiles."""
        alerts = []
        data = params.get("data", "")
        to = params.get("to", "")

        if not data or len(data) < 10:
            return alerts

        selector = data[:10] if data.startswith("0x") else "0x" + data[:8]

        # Unlimited approval
        if selector in ("0x095ea7b3", "0x39509351"):
            if "ffffffff" in data.lower():
                alerts.append(Alert(
                    profile="approval_phishing", heuristic="H1",
                    severity="high", confidence=0.95,
                    signal=f"Agent approving unlimited tokens to {to}",
                    recommendation="Limit approval to exact amount needed",
                ))

        return alerts

    def _check_signature(self, params: dict) -> list[Alert]:
        """Check signature against offchain_signature profile."""
        alerts = []
        typed_data = params.get("typed_data", params)

        primary_type = ""
        if isinstance(typed_data, dict):
            primary_type = typed_data.get("primaryType", "")

        if primary_type in ("PermitSingle", "PermitBatch"):
            alerts.append(Alert(
                profile="offchain_signature", heuristic="H1",
                severity="high", confidence=0.9,
                signal=f"Agent signing {primary_type} — authorizes token transfer off-chain",
                recommendation="Verify spender and amount. Set short expiration.",
            ))

        return alerts

    def _check_rpc_pattern(self) -> list[Alert]:
        """Check RPC query patterns against rpc_leakage profile."""
        alerts = []
        cutoff = time.time() - 300  # 5 min window
        recent = [q for q in self.rpc_log if q["time"] > cutoff]

        # Multiple balance checks. eth_getBalance can carry params as
        # either a list ([address, block]) or a dict ({"address": ...});
        # handle both shapes safely.
        balance_targets = set()
        for q in recent:
            if q["method"] == "eth_getBalance":
                p = q.get("params", None)
                addr = ""
                if isinstance(p, dict):
                    addr = p.get("address", "")
                elif isinstance(p, list) and p:
                    addr = p[0] if isinstance(p[0], str) else ""
                if addr:
                    balance_targets.add(addr)

        if len(balance_targets) > 3:
            alerts.append(Alert(
                profile="rpc_leakage", heuristic="H1",
                severity="high", confidence=0.8,
                signal=f"Agent checked balances for {len(balance_targets)} addresses — links them to same user",
                recommendation="Use local light client or batch through Tor",
            ))

        return alerts


# --- Demo ---

def demo():
    print("=" * 60)
    print("e_AI v2 AI Agent Guard Demo")
    print("=" * 60)

    guard = AgentGuard(profiles_dir="domains/")
    print(f"Loaded {len(guard.profiles)} profiles")

    # Simulate agent tools
    @guard.check
    def approve_token(token: str, spender: str, amount: str):
        print(f"  [Agent] Approving {amount} {token} to {spender}")
        return True

    @guard.check
    def check_balance(address: str):
        print(f"  [Agent] Checking balance of {address}")
        return 100.0

    @guard.check
    def execute_swap(token_in: str, token_out: str, amount: float):
        print(f"  [Agent] Swapping {amount} {token_in} → {token_out}")
        return True

    # Scenario 1: Agent tries unlimited approval
    print("\n--- Scenario 1: Agent approves unlimited tokens ---")
    try:
        approve_token("USDC", "0xunknown", "unlimited")
    except RuntimeError as e:
        print(f"  BLOCKED: {e}")

    # Scenario 2: Agent checks multiple balances (RPC leakage)
    print("\n--- Scenario 2: Agent checks 5 wallet balances ---")
    for addr in ["0xaaa", "0xbbb", "0xccc", "0xddd", "0xeee"]:
        check_balance(addr)

    # Scenario 3: Normal swap (should pass)
    print("\n--- Scenario 3: Agent executes normal swap ---")
    execute_swap("ETH", "USDC", 1.0)

    print("\n" + "=" * 60)


if __name__ == "__main__":
    demo()
